### PR TITLE
bootloader/many: rm ConfigFile, add Present for indicating presence of bloader

### PR DESCRIPTION
**Pull request by anonymouse64**
Now we have a specific method on the Bootloader interface which tells Find()
whether the specific bootloader is actually present on the current system. This
will simplify future bootloader implementations to be able to return errors when
identifying if a bootloader is present on the system is more complicated a
question than just if the config file for that bootloader exists.

This is required for the UC20 lk bootloader work, where the Present() implementation will be more complicated than just whether a file exists.

### CODE REVIEW

**Review by mvo5 — APPROVED** [review]
Thank you!

File: gadget/validate.go [hunk 1]
```diff
@@ -77,6 +77,10 @@ type ValidationConstraints struct {
 // are handled identically to ReadInfo(). Optionally takes additional validation
 // constraints, which for instance may only be known at run time,
 func Validate(gadgetSnapRootDir string, model Model, extra *ValidationConstraints) error {
+	// TODO: also validate that only one "<bl-name>.conf" file is in the root
```
Comment (bboozzoo) [inline gadget/validate.go:80]: On top of that, I think we should validate that the bootloader in gadget.yaml has a conf file inside. Ideally ForGadget could use that knowledge too.

File: bootloader/bootloader.go [hunk 2]
```diff
@@ -357,12 +366,11 @@ func ForGadget(gadgetDir, rootDir string, opts *Options) (Bootloader, error) {
 	if forcedBootloader != nil || forcedError != nil {
 		return forcedBootloader, forcedError
 	}
-	for _, blNew := range bootloaders {
-		bl := blNew(rootDir, opts)
-		markerConf := filepath.Join(gadgetDir, bl.Name()+".conf")
+	for name, blNew := range bootloaders {
+		markerConf := filepath.Join(gadgetDir, name+".conf")
```
Comment (bboozzoo) [inline bootloader/bootloader.go:370]: Hm I think I'd rather keep the previous way of creating the bootloader and then checking for existence of `bl.Name()+".conf"`, unless there's a reason to reverse the order of like it is now.

File: bootloader/bootloader.go [hunk 3]
```diff
@@ -291,9 +295,14 @@ func Find(rootdir string, opts *Options) (Bootloader, error) {
...
 	}
 
-	for _, blNew := range bootloaders {
+	// note that the order of this is not deterministic
+	for name, blNew := range bootloaders {
 		bl := blNew(rootdir, opts)
-		if osutil.FileExists(bl.ConfigFile()) {
+		present, err := bl.Present()
+		if err != nil {
+			return nil, fmt.Errorf("bootloader %q found but not usable: %v", name, err)
```
Comment (bboozzoo) [inline bootloader/bootloader.go:303]: this could be bl.Name() and then we don't need bootloaders to be a map[string]bootloaderNewFunc

**Review by pedronis — APPROVED** [review]
thank you

File: bootloader/bootloader.go (hunk 2 above)
Comment (bboozzoo) [reply bootloader/bootloader.go:370]: Thinking a bit more, what if we cold do:
```
blName, err := gadget.RequiredBootloader(gadgetDir)
..
blNew := bootloaders[blName]
..
bl := blNew(rootDir, opts)
bl.InstallableFor(gadgetDir) // checks presence of Name()+".conf" internally
```

Comment (bboozzoo) [reply bootloader/bootloader.go:370]: opened this for the gadget: https://github.com/snapcore/snapd/pull/9677

Comment (anonymouse64) [reply bootloader/bootloader.go]: this is now the old way of doing things again

File: bootloader/bootloader.go (hunk 3 above)
Comment (anonymouse64) [reply bootloader/bootloader.go]: this too

**Pull request merged**